### ISSUE TITLE
Annotations for adapters, listener hooks etc

### ISSUE THREAD

**Issue by lteacher** [lteacher/shellstone#13]
**Idea**
It would be nice to unify the annotations they should all be simple obvious and make clear the purpose.

Specifically I am hating the `DBAdapter` annotation right now, this is directly related to issue #12. It should be clear if something is an adapter, and the annotation still provides a nice way to plug a custom adapter in as a drop in replacement or as a completely new one. I think the adapter should be clear and the listener hook things should be also clear.

Perhaps...

``` dart
@Model(...) // I am pretty happy with this one
@Attr() // Again this makes enough sense
@Adapter() // I think this would be a nicer replacement for @DBAdapter
```

Finally the listener hooks I am not sure yet, ideally the whole framework will broadcast events around which can be hooked into. I don't _really_ dig the class based approach to doing that but it is very easy right now with the code, and also the annotation doesn't look too ridiculous.

This is one idea, that I am not exactly digging..

``` dart
@Listen(Adapter.configure)
doStuff(adapter) => 'stuff';
```

Also the issue with that is that it has become clear that sometimes for the events there is a need for a `pre` and `post` event. I think this one might be a bit later to get to and require having defined fully the broadcast events.

**Comment by lteacher** [comment]
Default is `pre` however has named constructor for `pre` and `post` events. `Adapter.configure` is a final event on the annotation?

``` dart
@Listen.post(Adapter.configure)
notifyEveryone() => 'It's all configured!';
```

**Comment by lteacher** [comment]
I'm going with a strategy that looks like:

``` dart
/* There is also Hook.pre and Listen.pre */
@Hook(Adapter.configure)
setCredentials(event) { }

@Listen(Adapter.configure)
logBeforeConfig(event) { }

@Listen.post(Adapter.configure)
logAfterConfig(event) { }
```

Where `@Listen` is a notified function that will be forgotten whilst `@Hook` is a function that can intercept during the event.

**Comment by lteacher** [comment]
Implemented in #17
